pkg/keyspace/tso_keyspace_group.go

```diff
@@ -16,24 +16,59 @@ package keyspace
 
 import (
 	"context"
+	"sync"
+	"time"
 
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/balancer"
+	"github.com/tikv/pd/pkg/mcs/discovery"
 	"github.com/tikv/pd/pkg/mcs/utils"
 	"github.com/tikv/pd/pkg/storage/endpoint"
 	"github.com/tikv/pd/pkg/storage/kv"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultBalancerPolicy = balancer.PolicyRoundRobin
+	allocNodeTimeout      = 1 * time.Second
+	allocNodeInterval     = 10 * time.Millisecond
 )
 
 // GroupManager is the manager of keyspace group related data.
 type GroupManager struct {
-	ctx context.Context
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
 	// store is the storage for keyspace group related information.
 	store endpoint.KeyspaceGroupStorage
+
+	client *clientv3.Client
+
+	// tsoServiceKey is the path of TSO service in etcd.
+	tsoServiceKey string
+	// tsoServiceEndKey is the end key of TSO service in etcd.
+	tsoServiceEndKey string
+
+	policy balancer.Policy
+
+	nodesBalancer balancer.Balancer[string]
```

Comment:
It seems that we just select nodes in the round-robin. We need two-levels scheduling here. The first level is to use the following policies to place keyspace groups according to their UserKind, within each UserKind then user round-robin. We probably can't skip these two-levels scheduling for now. + @rleungx, thoughts? 

1. The TSO system tries to evenly distribute the keyspace with the User Kind "business" or "enterprise", across different keyspace groups as much as possible.
2. The TSO system tries to evenly distribute the primary/secondaries of the keyspace groups with the User Kind "business" or "enterprise", across the TSO nodes as much as possible.
3. The TSO system tries to assign the 'basic' keyspace groups' primaries to as few TSO nodes as possible.

Reply:
So where will the correspondence between tso node and user kind be found?

Reply:
The balancer assigns and specifies.